controller/server/gui.ml

```diff
@@ -12,17 +12,38 @@ let resource_path end_path =
   (Sys.argv.(0) |> v |> parent) / ".." / "share" // end_path
   |> to_string
 
+
 (* Load a template file
 *)
 let template name =
   Fpath.(resource_path (v "template" / (name ^ ".mustache")))
   |> Util.read_from_file log_src
   >|= Mustache.of_string
 
+(* Load a partial file
+ * This had to be defined separately from the function loading templates,
+ * because its has to have the type string -> Mustache.t option (cannot use lwt).
+*)
+let load_partial name =
+   let load_file f =
+     let ic = open_in f in
+     let n = in_channel_length ic in
+     let s = Bytes.create n in
+     really_input ic s 0 n;
+     close_in ic;
+     Bytes.to_string s
+   in
+   let partial =
+     Fpath.(resource_path (v "template" / "partials" / (name ^ ".mustache")))
+     |> load_file
+     |> Mustache.of_string
+   in
+   Some partial
```

Comment:
I started implementing this, but I'd like to check-in with you to see if I'm on the right path:
https://github.com/krksgbr/playos/compare/improve-network-config-ui...krksgbr:network-config-preload-templates

Pre-loading the templates and partials requires us to thread them through the render functions. I feel like it complicates things a bit and I am not sure if it's a good trade-off. Do you know a better way of doing this?

Reply:
The implementation looks good.

If there is an error in a template, it seems to raise `Parse_error` (https://github.com/rgrinberg/ocaml-mustache/blob/master/lib/mustache.mli#L67). One advantage of this approach is that it would fail directly at startup, so we would catch it directly during tests. But one disadvantage is that we would have to make the controller to run nonetheless, because we don’t want to break the update mechanism because of a template error.

I have mixed feelings about this.

What is currently happening in case there is an error in a template ? Or if there is an error in a partial ?

Reply:
If there is a parse error either in a template or a partial, an exception is raised and the program fails.
I think we could catch the exception, log the error, and return a dummy string for the template or partial which would also make it easier to track down:
![image](https://user-images.githubusercontent.com/7548374/109507143-b4b34100-7a9e-11eb-9e96-27c18d6d0a8f.png)

This is even better than what we have now, because currently we don't see where the error is coming from:

![image](https://user-images.githubusercontent.com/7548374/109507304-e0cec200-7a9e-11eb-8bdb-96eb3c46319d.png)

Reply:
> I think we could catch the exception, log the error, and return a dummy string for the template or partial which would also make it easier to track down.

Nice!

> This is even better than what we have now, because currently we don't see where the error is coming from.

But is it also doable with on the fly loading ?

Reply:
Yes, it's possible to do it.
I'm a bit confused now however: do you suggest to continue with the idea of pre-loading the templates or should we keep things as they are now and load them on the fly?

Reply:
> I'm a bit confused now however: do you suggest to continue with the idea of pre-loading the templates or should we keep things as they are now and load them on the fly?

I don’t know, I suggest we make a call to discuss pick a solution.